interface/patient_file/summary/demographics.php

```diff
@@ -253,33 +253,28 @@ function editScripts(url) {
         __this.find("#backButton").css("display", "");
         __this.find("#addButton").css("display", "none");
 
-        var iam = top.tab_mode ? top.frames.editScripts : window[0];
-        iam.location.href = "<?php echo $GLOBALS['webroot']?>/controller.php?prescription&edit&id=&pid=<?php echo attr_url($pid);?>"
+        var iam = top.tab_mode ? top.frames.editScripts : window[1];
+        iam.location.href = "<?php echo $GLOBALS['webroot']?>/controller.php?prescription&edit&id=0&pid=<?php echo attr($pid);?>"
```

Comment:
better if:
```
iam.location.href = "<?php echo $GLOBALS['webroot']?>/controller.php?prescription&edit&id=0&pid=" + <?php echo js_url($pid);?>;
```

Reply:
Not sure why but this concerns me some. Are we sure we want double quotes on query components? i.e.
`iam.location.href = '/openemr/controller.php?prescription&edit&id=0&pid="44"'`
of course I had to wrap in single quotes to get to work with js_url. Most experience dev would get this knowing this is a json returned from js_url but this is a component and probably best to use attr_js:
`iam.location.href = '/openemr/controller.php?prescription&edit&id=0&pid=44'`
I think js_url will serve better to use on full url's.
I'll use attr_js here. thanks

Reply:
hi @sjpadgett 

To better explain:
js_url will do the following:
1. run urlencode(), which should be just done on parameters/variables (for example, if the parameter contains a & or other chars that break url, then it wont break url)
2. then runs json_encode, which will escape code in js code (which this is) and place quotes on the ends. Note that we are not placing quotes on the query component. The quotes are simply being used to concatenate a js string (ie. that is why the + sign is used)

Since we are in raw js, then no need for attr(). And in this case doing the attr_js here as doing will place escaped quotes in the pid parameter.

-brady

Reply:
Check this out to get an idea of how all these crazy functions are used:
https://www.open-emr.org/wiki/index.php/Codebase_Security#Cross-Scripting_Prevention
(looks like I need to add an example for js_url() use there :) )

Reply:
> iam.location.href = '/openemr/controller.php?prescription&edit&id=0&pid=44'
This is what the attr_js will result. It's the js_url that will result in double quotes for pid.

Reply:
also note I changed in my last commit

Reply:
First question to sort this out:
Why not running through urlencode() ?

Second question to sort this out:
For the json_encode() portion did you ensure that used string concatenation in js:
(note I messed up my example above :) )
```
iam.location.href = '<?php echo $GLOBALS['webroot']; ?>/controller.php?prescription&edit&id=0&pid=' + <?php echo js_url($pid); ?>;
```

Reply:
Looks like json_encode() may not place quotes around integers. I'm gonna do some low level testing on these functions for a bit. Will get back to you after I have a better handle on this stuff.

Reply:
testing pad to see why things are not going as I expect them to :)
```
iam.location.href = '<?php echo $GLOBALS['webroot']?>/controller.php?prescription&edit&id=0&pid=<?php $hey = 'hey'; echo attr_js($hey); ?>'
```
outputs:
```
iam.location.href = '/controller.php?prescription&edit&id=0&pid=&quot;hey&quot;'
```
BUT
```
iam.location.href = '<?php echo $GLOBALS['webroot']?>/controller.php?prescription&edit&id=0&pid=<?php $hey = 10; echo attr_js($hey); ?>'
```
outputs:
```
iam.location.href = '/controller.php?prescription&edit&id=0&pid=10'
```
So, the json_encode() function does not place quotes around integers, which makes sense since ok to do that in javascript.

NOW onto js_url when use string concatenation:
```
iam.location.href = '<?php echo $GLOBALS['webroot']?>/controller.php?prescription&edit&id=0&pid=' + <?php $hey = "hey"; echo js_url($hey); ?>;
```
outputs:
```
iam.location.href = '/controller.php?prescription&edit&id=0&pid=' + "hey";
```
and
```
iam.location.href = '<?php echo $GLOBALS['webroot']?>/controller.php?prescription&edit&id=0&pid=' + <?php $hey = 10; echo js_url($hey); ?>;
```
outputs:
```
iam.location.href = '/controller.php?prescription&edit&id=0&pid=' + "10";
```
This is really interesting, because if I use json_encode instead to do:
```
iam.location.href = '<?php echo $GLOBALS['webroot']?>/controller.php?prescription&edit&id=0&pid=' + <?php $hey = 10; echo js_encode($hey); ?>;
```
it also outputs:
```
iam.location.href = '/controller.php?prescription&edit&id=0&pid=' + "10";
```
This is super interesting, because it appears that when json_encode() operates on a integer, it must use context to decide whether to place quotes on the outside of it. And in the case here, attr_js() is not wrapping quotes around an integer, however it will break if not an integer.

Reply:
I'm sorry Brady but js_url() doesn't make sense to me. Why write a statement like:
`json_encode(urlencode(44))` 
so: `iam.location.href = '<?php echo $GLOBALS['webroot']?>/controller.php?prescription&list&id=<?php echo js_url(44); ?>'`
results in: `iam.location.href = '/openemr/controller.php?prescription&list&id="44"'`
which is really not a problem but where js_url can becomes misleading is when used like:
`js_url("&pid=44&id=$blah")` where one may expect `urlencode("&pid=44&id=$blah")` to be proper but then encoding again with json makes forming a full url problematic.
But we all know i'm no escaping wiz fur sure:)

Reply:
Lets just focus on urlencode first, which is separate from the XSS prevention.

urlencode is just basically used to sanitize GET parameters when building the path manually. For example, imagine if you build a path with a GET variable that contains a & characters. Since the & character is used to delimit GET parameters, this character in a parameter itself would break the entire request. urlencode with encode the & character so it doesn't break it. So, in your above example, would do:
```
"&pid=44&id=" . urlencode($blah)
```
On an interesting geeky escaping footnote. JS also has a function for this, which is encodeURIComponent(). If above were inside js code, then another way would be:
```
"&pid=44&id=" + encodeURIComponent("<php echo $blah; ?>")
```
Let me know if that makes sense or if still seems crazy

Reply:
yep, I think I was looking at urlencode wrongly from a php point of view and more towards ajax/JS encoding. I think my problem is that I don't look at escaping from the security aspects  but more towards the practical side. I can make any of it work in building the statements by dotting my T's and crossing my I's so, I leave it to you guys to keep me honest. I still don't like the idea of having double quotes on my query components. That's just me.

Reply:
k, going to assume we're still in business. Lets now focus on the second one above. Lets continue to pretend the block is within javascript code block. Lets change it to have php do the urlencode instead first:
```
iam.location.href =  "&pid=44&id=" + "<php echo urlencode($blah); ?>";
```
(note the + character and note the quotes)
NOW, lets work in json_encode for cross scripting protection:
```
iam.location.href = "&pid=44&id=" + <php echo json_encode(urlencode($blah)); ?>;
```
(note the + character and removal of the quotes)
Note this is basically the same as the js_url() function, which will look like so:
```
iam.location.href = "&pid=44&id=" + <php echo js_ur($blah); ?>;
```
of which the output will look something like:
```
iam.location.href = "&pid=44&id=" + "44";
```
yes, there are quotes there, but this it to build the js string. the actual iam.location.href value now equals &pid=44&id=44 (ie. no quotes)
(your mistake, which is producing the quotes in the value is that you are not using the + sign)

Reply:
I'm still with ya but just couldn't stay awake last night. I see what you're doing and I suppose I need to study up on security some. It just seems a lot to get my lowly pid to the world. I suppose it is easier than encoding on the javascript side.
My next PR i'll push a commit to change demographics to the js_url. Thanks for taking the time. Maybe some following this will get something from it.

Reply:
All variables get equal treatment. Even the lowly ones :)

Reply:
btw, don't worry about fixing it. I just brought in a PR that did this along with a couple other places in this script.